### PR TITLE
Add source files and examples

### PR DESCRIPTION
**Pull request by nomad605dis**

### CODE REVIEW

File: examples/MultipleOctofet/MultipleOctofet.ino [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// Two libraries have to be included
+// to work with the module. Download on
+// https://amperka.com/my/octofet
+#include <Octofet.h>
+#include <SPI.h>
+
+constexpr auto pinCS = 10;
```
Comment (AlexErAmp) [inline examples/MultipleOctofet/MultipleOctofet.ino:8]: Please add a comment that these expressions can be changed by the user.

File: examples/MultipleOctofet/MultipleOctofet.ino [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+void loop() {
+    // Set the "on" state for every switches of every device one by one
+    for (int t = 0; t < deviceCount; t++) {
+        for (int i = 0; i < 8; i++) {
+            octofet.digitalWrite(i, HIGH, t);
+            delay(1000);
+        }
+    }
+    // Similarly, set switches to "off" state
+    for (int t = 0; t < deviceCount; t++) {
```
Comment (AlexErAmp) [inline examples/MultipleOctofet/MultipleOctofet.ino:37]: It was possible not to do a separate cycle to turn off the switches. You could not do a separate cycle to turn off the keys. You just need to create the `state` variable, which you need to toogle after the first loop.

File: examples/SingleOctofet/SingleOctofet.ino [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// Two libraries have to be included
+// to work with the module. Download on
+// https://amperka.com/my/octofet
+#include <Octofet.h>
+#include <SPI.h>
+
+constexpr auto pinCS = 10;
```
Comment (AlexErAmp) [inline examples/SingleOctofet/SingleOctofet.ino:8]: Please add a comment that these expression can be changed by the user.

File: src/Octofet.cpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
+#include "octofet.h"
+
+Octofet::Octofet(uint8_t pinCS, uint8_t deviceCount)
+    : _spi(pinCS, deviceCount) {
+    _deviceCount = deviceCount;
```
Comment (AlexErAmp) [inline src/Octofet.cpp]: The variable `_deviceCount` can be transferred to the initialization list.

File: src/Octofet.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+Octofet::Octofet(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t deviceCount)
+    : _spi(pinCS, pinMOSI, pinSCK, deviceCount) {
+    _deviceCount = deviceCount;
```
Comment (AlexErAmp) [inline src/Octofet.cpp]: The variable `_deviceCount` can be transferred to the initialization list.

**Review by AlexErAmp — CHANGES_REQUESTED** [review]

File: examples/SingleOctofet/SingleOctofet.ino [hunk 6]
```diff
@@ -0,0 +1,45 @@
+// Two libraries have to be included
+// to work with the module. Download on
```
Comment (nkrkv) [inline examples/SingleOctofet/SingleOctofet.ino]:
```suggestion
// to work with the module. Download at
```

File: src/spi2parallel.cpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    _pinMOSI = pinMOSI;
+    _pinSCK = pinSCK;
+    _devicesCount = devicesCount;
+    _bitOrder = bitOrder;
+    _devicesChain = new uint8_t[_devicesCount - 1];
+    _hardwareSPI = false;
+}
+
+Spi2Parallel::~Spi2Parallel() {
+    if (_devicesChain != NULL) {
```
Comment (nkrkv) [inline src/spi2parallel.cpp]:
```suggestion
    if (_devicesChain != nullptr) {
```

File: src/spi2parallel.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+class Spi2Parallel {
+public:
+    Spi2Parallel(uint8_t pinCS, uint8_t devicesCount, uint8_t bitOrder = MSBFIRST);
+    Spi2Parallel(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t devicesCount, uint8_t bitOrder = MSBFIRST);
+    ~Spi2Parallel();
+    void begin();
+    void writeBit(uint8_t bit, bool value, uint8_t device);
+    void writeByte(uint8_t value, uint8_t device);
+    bool readBit(uint8_t bit, uint8_t device);
+    uint8_t readByte(uint8_t device);
```
Comment (nkrkv) [inline src/spi2parallel.h]:
```suggestion
    bool readBit(uint8_t bit, uint8_t device) const;
    uint8_t readByte(uint8_t device) const;
```

File: src/spi2parallel.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    void update();
+    uint8_t chainLength() { return _devicesCount; }
```
Comment (nkrkv) [inline src/spi2parallel.h]:
```suggestion
    uint8_t chainLength() const { return _devicesCount; }
```

File: src/Octofet.h [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+class Octofet {
+public:
+    Octofet(uint8_t pinCS, uint8_t deviceCount = 1);
+    Octofet(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t deviceCount = 1);
+    void begin();
+    void digitalWrite(uint8_t channel, bool value, uint8_t device = 0);
+    void digitalWrite8(uint8_t value, uint8_t device = 0);
+    bool getChannelState(uint8_t channel, uint8_t device = 0);
+    uint8_t getChannelState8(uint8_t device = 0);
```
Comment (nkrkv) [inline src/Octofet.h]:
```suggestion
    bool getChannelState(uint8_t channel, uint8_t device = 0) const;
    uint8_t getChannelState8(uint8_t device = 0) const;
```

File: examples/MultipleOctofet/MultipleOctofet.ino [hunk 11]
```diff
@@ -0,0 +1,40 @@
+// Two libraries have to be included
+// to work with the module. Download on
```
Comment (nkrkv) [inline examples/MultipleOctofet/MultipleOctofet.ino]:
```suggestion
// to work with the module. Download at
```

File: examples/MultipleOctofet/MultipleOctofet.ino [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+// constexpr auto pinSCK = 5;
+// Octofet octofet(pinCS, pinMOSI , pinSCK, deviceCount);
+
+void setup() {
+    // Start the board for communication
+    octofet.begin();
+}
+
+void loop() {
+    // Set the "on" state for every switches of every device one by one
```
Comment (nkrkv) [inline examples/MultipleOctofet/MultipleOctofet.ino]:
```suggestion
    // Set the "on" state for each switch of each device one by one
```

**Review by nkrkv — CHANGES_REQUESTED** [review]
:+1: :+1: 

A few boring points below:

- Search and replace “devicesCount” with “deviceCount”
- The same for “devicesChain” → “deviceChain”
- Add EOL at EOF where it is missing

File: examples/MultipleOctofet/MultipleOctofet.ino (hunk 2 above)
Comment (igor605ds) [reply examples/MultipleOctofet/MultipleOctofet.ino]: I think this is too difficult for beginners users.

Comment (nomad605dis) [reply examples/MultipleOctofet/MultipleOctofet.ino:37]: I think this is too difficult for beginners users.

File: examples/MultipleOctofet/MultipleOctofet.ino (hunk 11 above)
Comment (nomad605dis) [reply examples/MultipleOctofet/MultipleOctofet.ino]: Done

File: examples/MultipleOctofet/MultipleOctofet.ino (hunk 12 above)
Comment (nomad605dis) [reply examples/MultipleOctofet/MultipleOctofet.ino]: Done

File: examples/SingleOctofet/SingleOctofet.ino (hunk 6 above)
Comment (nomad605dis) [reply examples/SingleOctofet/SingleOctofet.ino]: Done

File: src/Octofet.cpp (hunk 4 above)
Comment (nomad605dis) [reply src/Octofet.cpp]: 👍

File: src/Octofet.cpp (hunk 5 above)
Comment (nomad605dis) [reply src/Octofet.cpp]: 👍

File: src/Octofet.h (hunk 10 above)
Comment (nomad605dis) [reply src/Octofet.h]: Done. Also, added the qualifier "const" to the cpp files.

File: src/spi2parallel.cpp (hunk 7 above)
Comment (nomad605dis) [reply src/spi2parallel.cpp]: Done

File: src/spi2parallel.h (hunk 8 above)
Comment (nomad605dis) [reply src/spi2parallel.h]: Done. Also, added the qualifier "const" to the cpp files.

File: src/spi2parallel.h (hunk 9 above)
Comment (nomad605dis) [reply src/spi2parallel.h]: Done

File: examples/MultipleOctofet/MultipleOctofet.ino (hunk 1 above)
Comment (nomad605dis) [reply examples/MultipleOctofet/MultipleOctofet.ino:8]: Done

File: examples/SingleOctofet/SingleOctofet.ino (hunk 3 above)
Comment (nomad605dis) [reply examples/SingleOctofet/SingleOctofet.ino:8]: Done

**Review by nomad605dis — COMMENTED** [review]
1. Fix lexical and stylistic mistakes.
2. Fix boring points from @nkrkv

File: examples/MultipleOctofet/MultipleOctofet.ino (hunk 11 above)
Comment (nkrkv) [reply examples/MultipleOctofet/MultipleOctofet.ino]: Not done. Still “Download on”. Should be “Download **at**”

File: examples/SingleOctofet/SingleOctofet.ino (hunk 6 above)
Comment (nkrkv) [reply examples/SingleOctofet/SingleOctofet.ino]: Ditto

File: src/Octofet.cpp [hunk 13]
```diff
@@ -0,0 +1,36 @@
+#include "octofet.h"
```
Comment (nkrkv) [inline src/Octofet.cpp]:
```suggestion
#include "Octofet.h"
```

Linux and macOS are case-sensitive. Please, verify the library compiles correctly on Linux.

Comment (nomad605dis) [reply src/Octofet.cpp]: Done. The library compiles and works correctly on Linux.

**Review by nomad605dis — COMMENTED** [review]
Fix names for Linux and macOS. The library compiles and works correctly on Linux.

File: library.properties [hunk 14]
```diff
@@ -0,0 +1,9 @@
+name=Octofet
+version=1.0.0
+author=Igor Dementiev <igor@amperka.ru>
+maintainer=Amperka <dev@amperka.com>
+sentence=Allows to control the power switches on board Octofet.
```
Comment (nkrkv) [inline library.properties]:
```suggestion
sentence=Amperka Octofet board (eight-channel power switch) interface library.
```

File: library.properties [hunk 15]
```diff
@@ -0,0 +1,9 @@
...
+paragraph=The library allows you to control devices through the output shift register (74HC595) via the hardware or software SPI interface.
```
Comment (nkrkv) [inline library.properties]:
```suggestion
paragraph=Control multiple high-power loads through the Amperka Octofet board using hardware or software SPI interface.
```

**Review by nkrkv — CHANGES_REQUESTED** [review]

File: library.properties (hunk 14 above)
Comment (nomad605dis) [reply library.properties]: Done

File: library.properties (hunk 15 above)
Comment (nomad605dis) [reply library.properties]: Done

**Review by nkrkv — APPROVED** [review]
:+1:

**Review by AlexErAmp — APPROVED** [review]
:1st_place_medal:

**Comment by nomad605dis** [comment]
@yurybotov and @Matthew3000 review please.

File: examples/MultipleOctofet/MultipleOctofet.ino [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+// Create an object of Octofet type
+// to communicate with Octofet board via hardware SPI
+Octofet octofet(pinCS, deviceCount);
+
+// If software SPI needed
+// add MOSI and SCK pins numbers, like this:
+// constexpr auto pinMOSI = 7;
+// constexpr auto pinSCK = 5;
+// any GPIO pins can be used as pinMOSI and pinSCK
+// Octofet octofet(pinCS, pinMOSI , pinSCK, deviceCount);
```
Comment (Matthew3000) [inline examples/MultipleOctofet/MultipleOctofet.ino]:
```suggestion
// Octofet octofet(pinCS, pinMOSI, pinSCK, deviceCount);
```

**Review by Matthew3000 — CHANGES_REQUESTED** [review]

File: examples/MultipleOctofet/MultipleOctofet.ino (hunk 16 above)
Comment (nomad605dis) [reply examples/MultipleOctofet/MultipleOctofet.ino]: Done.

**Review by Matthew3000 — APPROVED** [review]
Yo

File: src/spi2parallel.cpp [hunk 17]
```diff
@@ -0,0 +1,81 @@
+#include "spi2parallel.h"
+
+Spi2Parallel::Spi2Parallel(uint8_t pinCS, uint8_t deviceCount, uint8_t bitOrder)
+    : _pinCS(pinCS)
+    , _deviceCount(deviceCount)
+    , _bitOrder(bitOrder)
+    , _deviceChain(new uint8_t[_deviceCount - 1])
```
Comment (yurybotov) [inline src/spi2parallel.cpp]: _deviceChain(new uint8_t[_deviceCount - 1]) - if _deviceCount == 1 we will have array size == 0. Remove '-1'.

File: src/spi2parallel.cpp [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+    , _hardwareSPI(true) {
+}
+
+Spi2Parallel::Spi2Parallel(uint8_t pinCS, uint8_t pinMOSI, uint8_t pinSCK, uint8_t deviceCount, uint8_t bitOrder)
+    : _pinCS(pinCS)
+    , _pinMOSI(pinMOSI)
+    , _pinSCK(pinSCK)
+    , _deviceCount(deviceCount)
+    , _bitOrder(bitOrder)
+    , _deviceChain(new uint8_t[_deviceCount - 1])
```
Comment (yurybotov) [inline src/spi2parallel.cpp]: And here. _deviceChain(new uint8_t[_deviceCount - 1]) - if _deviceCount == 1 we will have array size == 0. Remove '-1'.

File: src/spi2parallel.cpp [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    if (_deviceChain != nullptr) {
+        delete[] _deviceChain;
+    }
+}
+
+void Spi2Parallel::begin() {
+    pinMode(_pinCS, OUTPUT);
+    digitalWrite(_pinCS, LOW);
+    if (_hardwareSPI) {
+        //SPI.beginTransaction(SPISettings(1000000, _bitOrder, SPI_MODE0));
```
Comment (yurybotov) [inline src/spi2parallel.cpp:31]: safe it for future. SPI.setBitOrder now is deprecated, but SPI.beginTransaction is not working correctly

File: examples/MultipleOctofet/MultipleOctofet.ino [hunk 20]
```diff
@@ -0,0 +1,43 @@
+// Two libraries have to be included
+// to work with the module. Download at
+// https://amperka.com/my/octofet
+#include <Octofet.h>
+#include <SPI.h>
+
+// any GPIO pin can be used as pinCS
+constexpr auto pinCS = 10;
```
Comment (yurybotov) [inline examples/MultipleOctofet/MultipleOctofet.ino:8]: Igor, is it worth it to use **constexpr** in the examples calculated for a simple arduino user, and even **auto** to the same. We will confuse them. This question is not so much for you as for Victor.

**Review by yurybotov — CHANGES_REQUESTED** [review]
Fix please array size...(look comments)

File: examples/MultipleOctofet/MultipleOctofet.ino (hunk 20 above)
Comment (nkrkv) [reply examples/MultipleOctofet/MultipleOctofet.ino:8]: We demonstrate good practice here. It is better than the traditional:

```cpp
#define PIN_CS 10
```

Let it be.

File: src/spi2parallel.cpp (hunk 17 above)
Comment (nkrkv) [reply src/spi2parallel.cpp]: :open_mouth:  Indeed! Good catch!

How could it even pass a test?!

File: examples/MultipleOctofet/MultipleOctofet.ino (hunk 20 above)
Comment (nomad605dis) [reply examples/MultipleOctofet/MultipleOctofet.ino:8]: Ok, leaved as is.

File: src/spi2parallel.cpp (hunk 17 above)
Comment (nomad605dis) [reply src/spi2parallel.cpp]: Oooppss, thanks.

Function `new  uint8_t [0 ]` create a pointer to memory with 0 size. 
Unfortunately, everything worked in the Arduino IDE.

File: src/spi2parallel.cpp (hunk 18 above)
Comment (nomad605dis) [reply src/spi2parallel.cpp]: Oooppss, thanks.

Function `new  uint8_t [0 ]` create a pointer to memory with 0 size. 
Unfortunately, everything worked in the Arduino IDE.

File: src/spi2parallel.cpp (hunk 19 above)
Comment (nomad605dis) [reply src/spi2parallel.cpp:31]: Ok, this is described in the description:
https://www.arduino.cc/en/Reference/SPISetBitOrder

**Review by nomad605dis — COMMENTED** [review]
Fixed size array for _deviceChain.

**Review by yurybotov — APPROVED** [review]
Zo gut

**Pull request merged**